dotnet/src/Connectors/Connectors.Ollama/Services/OllamaChatCompletionService.cs

```diff
@@ -61,55 +108,271 @@ public OllamaChatCompletionService(
+
+            var functionCallContent = new FunctionCallContent(
+                functionName: functionName.Name,
+                pluginName: functionName.PluginName,
+                id: Guid.NewGuid().ToString().Substring(0, 8),
+                arguments: kernelArguments)
+            {
+                InnerContent = toolCall,
+                Exception = exception
+            };
+
+            result.Add(functionCallContent);
         }
 
-        return [new ChatMessageContent(
-            role: authorRole ?? new(),
-            content: fullContent.ToString(),
-            modelId: modelId,
-            innerContent: innerContent)];
+        return result;
+    }
+
+    private static Dictionary<string, object?> GetChatCompletionMetadata(ChatDoneResponseStream doneChatUpdate)
+    {
+        return new Dictionary<string, object?>
+            {
+                // Necessary for the function telemetry 
+                { "Usage", new {
+                    PromptTokens = doneChatUpdate.PromptEvalCount,
+                    CompletionTokens = doneChatUpdate.EvalCount
+                }},
+            };
+    }
+
+    /// <summary>Checks if a tool call is for a function that was defined.</summary>
+    private static bool IsRequestableTool(IEnumerable<Tool>? tools, FunctionCallContent functionCallContent)
+    {
+        if (tools is null)
+        {
+            return false;
+        }
+
+        foreach (var tool in tools)
+        {
+            if (tool.Function?.Name is not null &&
+                string.Equals(tool.Function!.Name, FunctionName.ToFullyQualifiedName(functionCallContent.FunctionName, functionCallContent.PluginName, OllamaChatCompletionService.FunctionNameSeparator), StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /// <summary>
+    /// Start a chat completion activity for a given model.
+    /// The activity will be tagged with the a set of attributes specified by the semantic conventions.
+    /// </summary>
+    private Activity? StartCompletionActivity(ChatHistory chatHistory, PromptExecutionSettings settings)
+        => ModelDiagnostics.StartCompletionActivity(this._client.Config.Uri, this._client.Config.Model, ModelPlatform, chatHistory, settings);
+
+    /// <summary>
+    /// Tracks tooling updates from streaming responses.
+    /// </summary>
+    /// <param name="updates">The tool call updates to incorporate.</param>
+    /// <param name="toolCallIdsByIndex">Lazily-initialized dictionary mapping indices to IDs.</param>
+    /// <param name="functionNamesByIndex">Lazily-initialized dictionary mapping indices to names.</param>
+    /// <param name="functionArgumentBuildersByIndex">Lazily-initialized dictionary mapping indices to arguments.</param>
+    internal static void TrackStreamingToolingUpdate(
```

Comment:
Question: What is the usage of the method?

Reply:
0 😂😂
Thanks!